src/compiler/checker.ts

```diff
@@ -11079,26 +11098,23 @@ namespace ts {
             // However, it is possible that either candidatesOutArray was not passed in the first time,
             // or that a different candidatesOutArray was passed in. Therefore, we need to redo the work
             // to correctly fill the candidatesOutArray.
-            if (!links.resolvedSignature || candidatesOutArray) {
-                links.resolvedSignature = anySignature;
+            const cached = links.resolvedSignature;
+            if (cached && cached !== anySignature && !candidatesOutArray) {
+                return cached;
+            }
+            links.resolvedSignature = anySignature;
+            const result = resolveSignature(node, candidatesOutArray);
+            // If signature resolution originated in control flow type analysis (for example to compute the
+            // assigned type in a flow assignment) we don't cache the result as it may be based on temporary
+            // types from the control flow analysis.
+            links.resolvedSignature = inFlowCheck ? cached : result;

```

Comment:
I see, so if the call has never been processed yet, the value of `cached` will be undefined at this point. We will restore the undefined if we are in a flow analysis. So if we revisit the call later, it will look unprocessed.

One thing that occurs to me is the following. Suppose one of the call arguments is a lambda. In that case, we will type the lambda according to the first visit in the flow analysis. Any lambda parameters we type will stick, and they will permanently reflect the typing of the first pass. To piggy back on your example:
```ts
declare function foo<T, U>(s: T, func: (x: T) => U): U;

function f() {
    let x: string | number | boolean;
    x = "";
    while (cond) {
        x; // string | number
        x = foo(x, y => y.length); // Should be an error, string | number has no property length
        x; // number
    }
}
```
It is essentially the same issue that we see with overloads, where a lambda parameter is typed according to the first overload that binds it. In this case, instead of different overloads, it's different iterations of the loop that pass an argument of a different type.